### ISSUE TITLE
Possible memory leak in lumina-desktop

### ISSUE THREAD

**Issue by mike-pt** [trueos/lumina#266]
In recent builds (I'm not sure when it started) but I've noticed the desktop gets extremely slow and even crash at time...

At first this seemed to be related to using Chromium (granted its not very stable on FreeBSD) but It seems that's actually lumina that's eating a lot of RAM, even with no apps open:

```c
 ps axvm |head
  PID STAT      TIME  SL  RE PAGEIN     VSZ     RSS LIM TSIZ  %CPU %MEM COMMAND
 1416 S     10:28.02   0 127   3034 3331720 2989112   - 2324   0.0 74.9 /usr/local/bin/lumina-desktop
 1245 S      1:01.17   3 127  20685  186876   30872   - 2104   0.0  0.8 /usr/local/bin/X -nolisten tcp vt09 -auth /var/run/slim.
 1409 I      0:00.50 127 127    303   82180    7288   - 1224   0.0  0.2 /usr/local/bin/fluxbox -rc /home/miguelc/.config/lumina-
  718 Ss     0:00.04   3 127      0    9512    2656   - 1096   0.0  0.1 /sbin/devd
    1 SLs    0:00.06   3 127     51    5388     272   -  992   0.0  0.0 /sbin/init --
 4483 IWs    0:00.00 127 117      0   35644       0   -  684   0.0  0.0 -zsh (zsh)
47633 Ss     0:01.63   0   2      1   35660    4788   -  684   0.6  0.1 -zsh (zsh)
```

```c
last pid: 47860;  load averages:  0.14,  0.20,  0.22                                                     up 0+02:50:08  14:47:57
63 processes:  1 running, 60 sleeping, 2 zombie
CPU:  0.1% user,  0.0% nice,  0.1% system,  0.1% interrupt, 99.7% idle
Mem: 2865M Active, 175M Inact, 651M Wired, 70M Free
ARC: 286M Total, 165M MFU, 103M MRU, 402K Anon, 2561K Header, 15M Other
Swap: 2048M Total, 261M Used, 1787M Free, 12% Inuse

  PID USERNAME       THR PRI NICE   SIZE    RES STATE   C   TIME    WCPU COMMAND
 1416 miguelc          2  20    0  3464M  2861M select  2  11:15   0.02% lumina-desktop
```

Any idea of anything that might be causing this? What can I look at to find more info?

**Comment by beanpole135** [comment]
Which version of Lumina are you running? (Run `lumina-info` to see)
We are looking into this same thing on the recent TrueOS build from Git right now.

**Comment by WoefulDerelict** [comment]
While tracking master I've encountered the same issue. Before killing it this morning only a couple hours after rebooting lumina-desktop was reportedly using 10 GB of memory.

lumina-info: 1.0.1 (Git Revision: 0b86570)

**Comment by mike-pt** [comment]
I left the laptop running with no apps, lumina crashed, because it run out of memory:

```csh
swap_pager: out of swap space
swap_pager_getswapspace(16): failed
pid 1416 (lumina-desktop), uid 1001, was killed: out of swap space
```

--> *Version*: 1.0.1 (git rev: 79adf4f)

**Comment by mike-pt** [comment]
Note that this started before that build, I'm just not sure when :\

**Comment by WoefulDerelict** [comment]
After syncing with master this afternoon and observing lumina-desktop I saw its memory footprint bloom to ridiculous proportions within moments of launch and continue to grow to an astounding size without any interaction whatsoever. It continued to gobble up memory until I terminated the session. 

lumina-info: 1.0.1 (Git Revision: 088eebf)

**Comment by RodMyers** [comment]
had iXsystems look at my laptop with memory leak. 

if you have okular installed, temporarily remove it until a fix can be pushed

**Comment by mike-pt** [comment]
I don't... but even if it was the cause... how could the fact that a package is simply installed cause the leak if its not even running, (meaning would just make sure the process is not running fix the leak!?)

I'm more inclined to this being due to any other reason... or at most a mix.

**Comment by RodMyers** [comment]
it was the "start" of my memory leak. Handed over my laptop to KenMoore, and he narrowed it down to the actual cause. It's fixed, and should pushed out in a few days

**Comment by WoefulDerelict** [comment]
After syncing up to https://github.com/trueos/lumina/commit/e24b7b3f822a55aa8210ed21d50349147426f7a3 lumina-desktop's memory usage no longer baloons out of control after launch.

lumina-info: 1.0.1 (Git Revision: e24b7b3)

**Comment by beanpole135** [comment]
Yes, the memory usage will no longer keep growing anymore - but I have not quite finished the fix to free up memory from the data classes which are no longer used.

Background on the problem:
The new application registration watcher is a static instance of a class which creates/destroys small data structures for each application registration file on the system (keeping them within a fast-lookup data structure for usage later).  The problem is that the small data structures are not being completely destroyed when they are cleaned up from the internal table (probably because they are being treated as static themselves since the parent is static). I made sure that new entries in the table will no longer get for invalid app registrations on every scan (okular was installing ~10 invalid registrations alongside the ones that worked - which is why installing that package produced the problem on the systems I was testing), which stopped the increased memory usage over time - but removing okular and/or other packages will not currently free up the memory used for those registrations. I am still working on that and hope to have it completely working sometime soon.

**Comment by WoefulDerelict** [comment]
The band-aid is certainly appreciated and one looks forward to the fix proper. It wouldn't surprise me if you are correct and the children of your static class were also being treated as static preventing the memory from being freed.

**Comment by beanpole135** [comment]
Ok, I think this issue has officially been squashed now.
I took some extra time and cleaned up the entire `*.desktop` file support system in the LuminaXDG.[cpp/h] files, resulting in a much cleaner implementation of the XDG support protocols and which also gets rid of most of the static variables within the library class.

I think I also discovered another small memory leak within the start menu itself, but it is trivial at the moment and only gets "triggered" when you reload the applications menu (such as changing the categorization option). So you might use an extra few MB of memory if you cycle that app menu a bunch of times (which I was doing when performing some tests on the system) - but that is an edge case which most user's will never run into.
